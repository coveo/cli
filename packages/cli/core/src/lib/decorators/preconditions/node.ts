import {getBinVersionPrecondition} from './binPreconditionsFactory.js';

export const IsNodeVersionInRange = getBinVersionPrecondition('node', {
  prettyName: 'Node.js',
});
