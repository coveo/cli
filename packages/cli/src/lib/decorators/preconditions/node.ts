import {constants} from 'os';
import * as dedent from 'dedent';

import type Command from '@oclif/command';
import {valid as isVersionValid, lt as isVersionLessThan} from 'semver';

import {SpawnProcessOutput, spawnProcessOutput} from '../../utils/process';

const linkToReadme =
  'https://github.com/coveo/cli/wiki/Node.js,-NPM-and-NPX-requirements';

export function IsNodeVersionAbove(requiredVersion: string) {
  return async function (target: Command) {
    if (!isVersionValid(requiredVersion)) {
      target.warn(dedent`
        Required version invalid: "${requiredVersion}".
        Please report this error to Coveo: https://github.com/coveo/cli/issues/new
      `);
      return false;
    }

    const output = await spawnProcessOutput('node', ['--version']);

    if (isMissingExecutable(output)) {
      target.warn(`${target.id} requires Node.js to run.`);
      warnHowToInstallNode(target);
      return false;
    }

    if (output.stderr || output.exitCode !== 0) {
      target.warn(dedent`
        ${target.id} requires a valid Node.js installation to run.
        An unknown error happened while trying to determine your node version with node --version
        ${output.stderr}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    if (isVersionLessThan(output.stdout, requiredVersion)) {
      target.warn(dedent`
        ${target.id} needs a Node.js version greater than ${requiredVersion}
        Version detected: ${output.stdout}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    return true;
  };
}

export function warnHowToInstallNode(target: Command) {
  target.warn(dedent`
    Please visit ${linkToReadme} for more detailed installation information.
  `);
}

export function isMissingExecutable(output: SpawnProcessOutput) {
  return (
    output.exitCode !== null &&
    Math.abs(output.exitCode) === constants.errno.ENOENT
  );
}
