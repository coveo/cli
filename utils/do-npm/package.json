{
  "name": "@coveo/do-npm",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsc"
  },
  "devDependencies": {
    "@types/npmcli__arborist": "^5.6.1",
    "@types/npmcli__package-json": "^2.0.0",
    "@types/tmp": "^0.2.3",
    "@verdaccio/types": "^10.7.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@npmcli/arborist": "^6.2.0",
    "tmp": "^0.2.1",
    "npm": "^9.4.0"
  }
}
