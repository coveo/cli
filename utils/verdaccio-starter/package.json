{
  "name": "@coveo/verdaccio-starter",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsc"
  },
  "devDependencies": {
    "@types/npmcli__arborist": "^5.6.1",
    "@types/npmcli__package-json": "^2.0.0",
    "@types/tmp": "^0.2.3",
    "@verdaccio/types": "^10.7.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@coveo/do-npm": "1.0.0",
    "@npmcli/arborist": "^6.2.0",
    "get-port": "6.1.2",
    "tmp": "^0.2.1",
    "verdaccio": "5.24.1"
  }
}
