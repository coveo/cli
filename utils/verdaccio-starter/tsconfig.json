{
  "compilerOptions": {
    "outDir": "dist",
    "skipLibCheck": true,
    "declaration": true,
    "declarationDir": "dist",
    "module": "Node16",
    "moduleResolution": "node16",
    "target": "ESNext",
    "sourceMap": true
  },
  "exclude": ["node_modules"],
  "include": ["src"]
}
