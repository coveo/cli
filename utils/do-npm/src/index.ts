import {
  spawnSync,
  spawn,
  SpawnOptions,
  SpawnSyncOptions,
} from 'node:child_process';
import {dirname, join} from 'node:path';
import {createRequire} from 'node:module';

const require = createRequire(import.meta.url);
const npmCliPath = join(dirname(require.resolve('npm')), 'bin', 'npm-cli.js');

export const npmSync = (args: string[], opts?: SpawnSyncOptions) =>
  spawnSync('node', [npmCliPath, ...args], opts);
export const npmAsync = (args: string[], opts?: SpawnOptions) =>
  spawn('node', [npmCliPath, ...args], opts);
