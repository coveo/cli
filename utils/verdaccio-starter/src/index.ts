import Arborist from '@npmcli/arborist';
import type {Content as PackageJson} from '@npmcli/package-json';
import {ConfigRuntime} from '@verdaccio/types';
import {ChildProcess, fork} from 'node:child_process';
import {writeFileSync, mkdirSync} from 'node:fs';
import {join, dirname} from 'node:path';
import {createRequire} from 'node:module';
import {dirSync, tmpNameSync} from 'tmp';
import {npmSync as npm} from '@coveo/do-npm';
import getPort from 'get-port';
const require = createRequire(import.meta.url);

export async function startVerdaccio(
  packagesToProxy: string[] | string,
  includeLocalDependent = true,
  publishProxiedPackages = true
): Promise<{verdaccioProcess: ChildProcess; verdaccioUrl: string}> {
  if (typeof packagesToProxy === 'string') {
    return startVerdaccio(
      [packagesToProxy],
      includeLocalDependent,
      publishProxiedPackages
    );
  }
  const computedPackagesToProxy = new Map<string, string[]>();

  if (!includeLocalDependent) {
    packagesToProxy.forEach((packageName) =>
      computedPackagesToProxy.set(packageName, [])
    );
  } else {
    await computePackagesToProxy(
      Array.from(packagesToProxy),
      computedPackagesToProxy
    );
  }
  const port = await getPort();
  const configPath = computeVerdaccioConfig(computedPackagesToProxy.keys());
  const verdaccioProcess = await runVerdaccio(configPath, port);
  const verdaccioUrl = `http://localhost:${port}`;
  if (publishProxiedPackages) {
    const npmrcPath = fakeNpmAuth(port);
    doPublishProxiedPackages(computedPackagesToProxy, verdaccioUrl, npmrcPath);
  }

  return {verdaccioProcess, verdaccioUrl};
}

async function computePackagesToProxy(
  packagesToProxy: string[],
  computedPackagesToProxy: Map<string, string[]>
) {
  const workspacePackages: Map<string, PackageJson> =
    await getWorkspacePackages();
  while (packagesToProxy.length > 0) {
    const currentPackage = packagesToProxy.pop()!;
    if (computedPackagesToProxy.has(currentPackage)) {
      continue;
    }
    if (!workspacePackages.has(currentPackage)) {
      continue;
    }
    const currentPackageJson = workspacePackages.get(currentPackage);
    const dependencies = [
      ...Object.keys(currentPackageJson?.dependencies ?? {}),
      ...Object.keys(currentPackageJson?.peerDependencies ?? {}),
    ].filter((dep) => dep && workspacePackages.has(dep));

    computedPackagesToProxy.set(currentPackage, dependencies);
    packagesToProxy.push(...dependencies);
  }
}

function runVerdaccio(
  configPath: string,
  verdaccioPort: number
): Promise<ChildProcess> {
  return new Promise((resolve, reject) => {
    const childFork = fork(
      require.resolve('verdaccio/bin/verdaccio'),
      ['--config', configPath, '--listen', verdaccioPort.toString()],
      {stdio: 'ignore'}
    );
    childFork.on('message', (msg: {verdaccio_started: boolean}) => {
      if (msg.verdaccio_started) {
        resolve(childFork);
      }
    });
    childFork.on('error', (err: any) => reject([err]));
    childFork.on('disconnect', (err: any) => reject([err]));
  });
}

async function getWorkspacePackages() {
  const workspacePackages = new Map<string, PackageJson>();
  const arb = new Arborist({path: process.env['npm_config_local_prefix']});
  await arb.loadActual();
  for (const child of arb.actualTree.fsChildren) {
    if (!child.package.name) {
      continue;
    }
    workspacePackages.set(child.package.name, child.package);
  }
  return workspacePackages;
}

function computeVerdaccioConfig(packagesToProxy: Iterable<string>) {
  const tmpDir = dirSync({unsafeCleanup: true});
  const storagePath = join(tmpDir.name, 'storage');
  mkdirSync(storagePath);
  const packages = getPackagesConfig(packagesToProxy);
  const config: ConfigRuntime = {
    self_path: 'foo',
    security: {
      api: {legacy: true},
      web: {
        sign: {ignoreExpiration: true},
        verify: {ignoreExpiration: true},
      },
    },
    uplinks: {
      npmjs: {
        url: 'https://registry.npmjs.org/',
        cache: false,
      },
    },
    packages,
    storage: storagePath,
  };
  const tmpFile = join(tmpDir.name, 'config.json');
  writeFileSync(tmpFile, JSON.stringify(config));
  return tmpFile;
}

function doPublishProxiedPackages(
  computedPackagesToProxy: Map<string, string[]>,
  verdaccioUrl: string,
  npmConfigPath: string
) {
  const computedPackagesToProxyArray = Array.from(computedPackagesToProxy);
  const visitedSinceLastPublish = [];
  while (computedPackagesToProxyArray.length > 0) {
    if (
      visitedSinceLastPublish.length === computedPackagesToProxyArray.length
    ) {
      throw new Error(
        `Error while preping verdaccio probably due to circular dependencies
      - visitedSinceLastPublish: ${JSON.stringify(visitedSinceLastPublish)}
      - computedPackageToProxyArray: ${JSON.stringify(
        computedPackagesToProxyArray
      )}`
      );
    }

    const currentPackage = computedPackagesToProxyArray.shift();
    // If the package has some unpublished local deps, push it back into the list and mark it as visited
    if (
      currentPackage[1].some((dependencyName) =>
        computedPackagesToProxyArray.some(
          (packageEntryYetToPublish) =>
            packageEntryYetToPublish[0] === dependencyName
        )
      )
    ) {
      visitedSinceLastPublish.includes(currentPackage[0]) ||
        visitedSinceLastPublish.push(currentPackage[0]);
      computedPackagesToProxyArray.push(currentPackage);
      continue;
    }

    const packageDirectory = dirname(
      require.resolve(join(currentPackage[0], 'package.json'))
    );
    npm(['publish'], {
      env: {
        ...process.env,
        npm_config_registry: verdaccioUrl,
        npm_config_userconfig: npmConfigPath,
      },
      stdio: 'ignore',
      cwd: packageDirectory,
    });

    visitedSinceLastPublish.length = 0;
  }
  // San-check: we should not have any visits on the last publish.
  if (visitedSinceLastPublish.length) {
    throw new Error(
      `Error while preping verdaccio probably due to circular dependencies
    - visitedSinceLastPublish: ${JSON.stringify(visitedSinceLastPublish)}
    - computedPackageToProxy: ${JSON.stringify(computedPackagesToProxy)}`
    );
  }
}

function getPackagesConfig(packagesToProxy: Iterable<string>) {
  const packages = {};
  for (const packageToProxy of packagesToProxy) {
    Object.assign(packages, {
      [packageToProxy]: {
        access: ['$all', '$anonymous'],
        publish: ['$all', '$anonymous'],
      },
    });
  }
  packages['**'] = {
    access: ['$all', '$anonymous'],
    publish: ['$all', '$anonymous'],
    proxy: ['npmjs'],
  };
  return packages;
}

function fakeNpmAuth(verdaccioPort: number) {
  const tmpNpmRc = tmpNameSync();
  writeFileSync(
    tmpNpmRc,
    `//localhost:${verdaccioPort}/:_authToken = OH_WOW_LOOK_A_SECRET`
  );
  return tmpNpmRc;
}
