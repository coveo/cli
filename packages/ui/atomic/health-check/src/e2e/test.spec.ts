import stripAnsi from 'strip-ansi';
import {ChildProcess} from 'node:child_process';
import {join} from 'node:path';
import {mkdirSync, readFileSync, writeFileSync} from 'node:fs';
import {npmSync} from '@coveo/do-npm';
import {startVerdaccio} from '@coveo/verdaccio-starter';
import {hashElement} from 'folder-hash';
import {DirResult, dirSync} from 'tmp';
import treeKill from 'tree-kill';

const getJsonPackagePath = (packageName: string, testDir: string) =>
  join(
    testDir,
    'src',
    'components',
    packageName.replace(/@coveo\//, ''),
    'package.json'
  );

const getComponentJsonPackage = (jsonPackagePath: string) => {
  const json = readFileSync(jsonPackagePath).toString();
  return JSON.parse(json);
};

const overwriteJsonPackage = (jsonPackagePath: string, json: {}) => {
  writeFileSync(jsonPackagePath, JSON.stringify(json, null, 2));
};

/**
 * Add required properties so health check assertions are met
 */
const addMissingPropertiesToJsonPackage = (json: {}) => {
  return {
    ...json,
    description:
      'Custom Atomic component for E2E testing: This is component should pass all Health checks',
    homepage: 'https://my-custom-atomic-component-source-code.com',
  };
};

const publish = (pkgName: string, cwd: string) => {
  return npmSync(['publish', '--dry-run', '-w', pkgName], {
    cwd,
  });
};

const ATOMIC_PROJECT_NAMES = [
  '@coveo/create-atomic-component',
  '@coveo/create-atomic-result-component',
];
const PACKAGE_NAME_HEALTH = '@coveo/atomic-component-health-check';

describe(PACKAGE_NAME_HEALTH, () => {
  let verdaccioProcess: ChildProcess;
  let tempDirectory: DirResult;
  let testDirectory: string;
  let npmCache: string;
  let verdaccioUrl: string;

  beforeAll(async () => {
    ({verdaccioUrl, verdaccioProcess} = await startVerdaccio([
      PACKAGE_NAME_HEALTH,
      ...ATOMIC_PROJECT_NAMES,
    ]));
    tempDirectory = dirSync({unsafeCleanup: true, keep: true});
    npmCache = join(tempDirectory.name, 'npm-cache');
    mkdirSync(npmCache);
  });

  afterAll(async () => {
    verdaccioProcess.pid && (await treeKill(verdaccioProcess.pid));
    console.log(tempDirectory.name);
    tempDirectory.removeCallback();
  });

  describe.each(ATOMIC_PROJECT_NAMES)('%s', (atomicProjectName) => {
    const args = ['valid-component-name'];
    const pkgName = 'valid-component-name';

    beforeAll(() => {
      testDirectory = join(tempDirectory.name, atomicProjectName);
      mkdirSync(testDirectory, {recursive: true});
    });

    it('should initialize a base project and a component', async () => {
      npmSync(['init', atomicProjectName.replace('/create-', '/'), ...args], {
        env: {
          ...process.env,
          npm_config_registry: verdaccioUrl,
          npm_config_cache: npmCache,
        },
        cwd: testDirectory,
      });

      expect(
        await hashElement(testDirectory, {
          folders: {
            exclude: ['**node_modules', 'dist'],
            ignoreRootName: true,
            ignoreBasename: true,
          },
          files: {
            include: ['*'],
            exclude: ['**package-lock.json', 'stencil-docs.json'],
            ignoreRootName: true,
            ignoreBasename: true,
          },
        })
      ).toMatchSnapshot();
    });

    it('should be able to install all deps without issues', () => {
      expect(
        npmSync(['install'], {
          cwd: testDirectory,
          env: {
            ...process.env,
            npm_config_registry: verdaccioUrl,
            npm_config_cache: npmCache,
          },
        }).status
      ).toBe(0);
    });

    it('should be able to build without issues', () => {
      expect(
        npmSync(['run', 'build', '-w', pkgName], {
          cwd: testDirectory,
        }).status
      ).toBe(0);
    });

    describe('health-check', () => {
      beforeAll(() => {
        const jsonPackagePath = getJsonPackagePath(pkgName, testDirectory);
        const initialJson = getComponentJsonPackage(jsonPackagePath);
        const updatedJson = addMissingPropertiesToJsonPackage(initialJson);
        overwriteJsonPackage(jsonPackagePath, updatedJson);
      });

      it('should be able to pass health checks', () => {
        const {stdout} = publish(pkgName, testDirectory);
        const message = stripAnsi(stdout.toString()).replace(/\n/g, '');
        expect(message).not.toContain('✖');
        expect(message).toContain('✔ Readme file');
        expect(message).toContain('✔ Component name');
        expect(message).toContain('✔ Required properties in package.json');
      });

      it('should be able to publish without issues', () => {
        expect(publish(pkgName, testDirectory).status).toBe(0);
      });
    });
  });
});
