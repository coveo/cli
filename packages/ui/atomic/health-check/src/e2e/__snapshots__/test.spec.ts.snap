// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@coveo/atomic-component-health-check @coveo/create-atomic-component should initialize a base project and a component 1`] = `
HashedFolder {
  "children": [
    HashedFile {
      "hash": "oNYr5PU1qGjrnrQmmNCEcfYvNgM=",
      "name": "package.json",
    },
    HashedFile {
      "hash": "dk1NzEu4Fb2ZY6TKzt21xMFvc2c=",
      "name": "readme.md",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "KemKxsZLHiHNylz5a6Z2IK7PwBw=",
                  "name": "package.json",
                },
                HashedFolder {
                  "children": [
                    HashedFile {
                      "hash": "XemX7AoOihA4SgQsvWtcStjeiGo=",
                      "name": "valid-component-name.css",
                    },
                    HashedFile {
                      "hash": "mSEayrhdQLFyjWOgicX78wD8seE=",
                      "name": "valid-component-name.tsx",
                    },
                  ],
                  "hash": "TLrPzQ215bJQgHPDj+8uidPXAqo=",
                  "name": "src",
                },
                HashedFile {
                  "hash": "DNTPZTeWJfCjTGqD+edK5gP5vUM=",
                  "name": "stencil.config.ts",
                },
                HashedFile {
                  "hash": "yMxJnSeOFw52ObOJrKQEh86v+HM=",
                  "name": "tsconfig.json",
                },
              ],
              "hash": "MdfhPw3BWNSYK2+cqb5JCBKTSpI=",
              "name": "valid-component-name",
            },
          ],
          "hash": "QXq8wEsNM1ZMS7Ou+h2wmWBghhs=",
          "name": "components",
        },
        HashedFile {
          "hash": "YVuwXfIZfjfeVyLQ6cMy3kUqC3A=",
          "name": "html.d.ts",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "L08E7dMK5e0wD5P8meKHfkDDa3o=",
              "name": "index.css",
            },
            HashedFile {
              "hash": "2XR/MKCv0fo3uRpZRiTyOBoHk7s=",
              "name": "index.html",
            },
            HashedFile {
              "hash": "ZRD3OAMjlKnlu7Yh3iQv3gCcG8c=",
              "name": "index.ts",
            },
          ],
          "hash": "V7y/8CwkVmHHCoFc1OqeZ9uA7Uo=",
          "name": "pages",
        },
      ],
      "hash": "78a7AxTS7XRN0oh0UpSs49JxK0Q=",
      "name": "src",
    },
    HashedFile {
      "hash": "BVZfiCzL+2kulE12K4HLAIl3kJ8=",
      "name": "stencil.config.ts",
    },
    HashedFile {
      "hash": "yMxJnSeOFw52ObOJrKQEh86v+HM=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "X21xbFQ7F4yHHuAr6nxokxpnb7I=",
  "name": "create-atomic-component",
}
`;

exports[`@coveo/atomic-component-health-check @coveo/create-atomic-result-component should initialize a base project and a component 1`] = `
HashedFolder {
  "children": [
    HashedFile {
      "hash": "oNYr5PU1qGjrnrQmmNCEcfYvNgM=",
      "name": "package.json",
    },
    HashedFile {
      "hash": "dk1NzEu4Fb2ZY6TKzt21xMFvc2c=",
      "name": "readme.md",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "NMHnNJYUQzGCOq3aSmWl3tNp4fI=",
                  "name": "package.json",
                },
                HashedFolder {
                  "children": [
                    HashedFile {
                      "hash": "0kZk+nhPZP6DNXEt1ds/y13pPn4=",
                      "name": "valid-component-name.css",
                    },
                    HashedFile {
                      "hash": "ShFXwot8XbCjD/A7x/1LQ0+hYI4=",
                      "name": "valid-component-name.tsx",
                    },
                  ],
                  "hash": "pFuu6S9wb+y09faE9Lxub6t+cfY=",
                  "name": "src",
                },
                HashedFile {
                  "hash": "DNTPZTeWJfCjTGqD+edK5gP5vUM=",
                  "name": "stencil.config.ts",
                },
                HashedFile {
                  "hash": "yMxJnSeOFw52ObOJrKQEh86v+HM=",
                  "name": "tsconfig.json",
                },
              ],
              "hash": "evmFecMLmIelFBXZvhawgWv8oWg=",
              "name": "valid-component-name",
            },
          ],
          "hash": "TcERGSTeYN1rbMN/FyAlrUBM5QU=",
          "name": "components",
        },
        HashedFile {
          "hash": "YVuwXfIZfjfeVyLQ6cMy3kUqC3A=",
          "name": "html.d.ts",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "L08E7dMK5e0wD5P8meKHfkDDa3o=",
              "name": "index.css",
            },
            HashedFile {
              "hash": "2XR/MKCv0fo3uRpZRiTyOBoHk7s=",
              "name": "index.html",
            },
            HashedFile {
              "hash": "ZRD3OAMjlKnlu7Yh3iQv3gCcG8c=",
              "name": "index.ts",
            },
          ],
          "hash": "V7y/8CwkVmHHCoFc1OqeZ9uA7Uo=",
          "name": "pages",
        },
      ],
      "hash": "WH4OEE8ocXd+I5wm1fVgvdhN+BM=",
      "name": "src",
    },
    HashedFile {
      "hash": "BVZfiCzL+2kulE12K4HLAIl3kJ8=",
      "name": "stencil.config.ts",
    },
    HashedFile {
      "hash": "yMxJnSeOFw52ObOJrKQEh86v+HM=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "H++4r0vIdCS/bp8KbwVzyRBbvdY=",
  "name": "create-atomic-result-component",
}
`;
